tests/core/test_signal.py

```diff
@@ -81,6 +83,81 @@ async def test_signal_connect_fails_if_different_backend_but_same_by_value():
     await sim_signal.connect(mock=False, backend=initial_backend)
 
 
+async def test_signal_connect_fails_if_different_backend_but_same_by_type():
+    signal = Signal()
+    connect_time_backend = MockSignalBackend(int)
+    await signal.connect(backend=connect_time_backend)  # works
+    with pytest.raises(ValueError):
+        await signal.connect(
+            backend=MockSignalBackend(int)
+        )  # fails since backend is different
+
+    # forces reconnect as per , but passes since the backend is the same
+    await signal.connect(backend=connect_time_backend)
+
+
+async def test_signal_connects_to_previous_backend(caplog):
+    caplog.set_level(logging.DEBUG)
+    int_mock_backend = MockSignalBackend(int)
+    original_connect = int_mock_backend.connect
+
+    async def new_connect(timeout=1):
+        await asyncio.sleep(0.1)
+        await original_connect(timeout=timeout)
+
+    int_mock_backend.connect = new_connect
+    signal = Signal(int_mock_backend)
+    await signal.connect()
+    assert signal._backend.datatype == int
+    await signal.connect()
+    response = f"Reusing previous connection to {signal.source}"
+    assert response in caplog.text
+
+
+async def test_signal_connects_with_force_reconnect(caplog):
+    caplog.set_level(logging.DEBUG)
+    signal = Signal(MockSignalBackend(int))
+    await signal.connect()
+    assert signal._backend.datatype == int
+    await signal.connect(force_reconnect=True)
+    response = f"Connecting to {signal.source}"
+    assert response in caplog.text
+    assert "Reusing previous connection to" not in caplog.text
+
+
+@pytest.mark.parametrize(
+    "first, second",
+    [(True, False), (True, False)],
+)
+async def test_rejects_reconnect_when_connects_have_diff_mock_status(
+    caplog, first, second
+):
+    caplog.set_level(logging.DEBUG)
+    signal = Signal(MockSignalBackend(int))
+    await signal.connect(mock=first)
+    assert signal._backend.datatype == int
+    with pytest.raises(RuntimeError) as exc:
+        await signal.connect(mock=second)
+
+    assert f"`connect(mock={second})` called on a function where the previous " in str(
+        exc
+    )
+
+    response = f"Connecting to {signal.source}"
+    assert response in caplog.text
+
+
+async def test_signal_lazily_connects(RE):
+    mock_signal_rw = soft_signal_rw(int, 0, name="mock_signal")
+    await mock_signal_rw.connect(mock=False)
+    RE(ensure_connected(mock_signal_rw, mock=True))
+    assert (
+        mock_signal_rw._connect_task
+        and mock_signal_rw._connect_task.done()
+        and not mock_signal_rw._connect_task.exception()
+    )
```

Comment:
that test is the only one that uses ensure_connected, but seems to pass even if I change is as much as here: 
![image](https://github.com/bluesky/ophyd-async/assets/56644812/ba3bd1e3-6c26-4929-ab23-e174a6f7f47e)

Reply:
https://github.com/bluesky/ophyd-async/pull/368#discussion_r1652513830

Same as above comment. This will work because it's a soft_signal, it shouldn't work on the first connect.

Reply:
ok, I see now - so here we test a scenario where the device rejects at first but later accepts. 

I think the best way to do this is to have a mock device that has its own connect with this 'fail on first connect' functionality, to mimic a real life device that misbehaves when we try to connect to it.  So a similar class to this one:

![image](https://github.com/bluesky/ophyd-async/assets/56644812/9592913f-9f92-454c-8dd5-457d5207d2c6)

Reply:
Do it by swapping out a `MockSignalBackend`'s connect. That way you'll still be using the regular connect method/functionality of Device and Signal.